compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
         return tileSize > 1 && tileSize < targetVectorSize;
       })) {
-    return failure();
+    return;
   }
 
-  // Target 16x16 tile sizes if there are AVX512 features and all the tile sizes
-  // are greater than or equal to 16.
+  // Target 16x16 tile sizes if there are AVX512 features and all the tile
+  // sizes are greater than or equal to 16.
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(entryPointFn);
   if (hasAVX512fFeature(targetAttr) &&
-      llvm::all_of(distConfig.minTileSizes, [](int64_t tileSize) {
+      llvm::all_of(minTileSizes, [](int64_t tileSize) {
         return tileSize == 1 || tileSize >= 16;
       })) {
     targetVectorSize = 16;
   }
 
   // Replace dims to be vectorized with the new tile sizes.
   std::replace_if(
-      distConfig.minTileSizes.begin(), distConfig.minTileSizes.end(),
+      minTileSizes.begin(), minTileSizes.end(),
       [](int64_t tileSize) { return tileSize > 1; }, targetVectorSize);
 
+  sizes = minTileSizes;
+  scalableSizeFlags = SmallVector<bool>(sizes.size(), false);
+}
+
+/// Utility to compute the tile sizes of transposes for AArch64 SME. Unlike
+/// other targets, the tile sizes picked here must exactly match multiples of
+/// the SME hardware virtual tiles, as there is currently no support for
+/// lowering non-standard shapes.
+static void
+getTransposeAArch64SMEVectorSizes(linalg::GenericOp genericOp,
+                                  SmallVectorImpl<int64_t> &sizes,
+                                  SmallVectorImpl<bool> &scalableSizeFlags) {
+  if (!transposeLoweringPreconditionAArch64SME(genericOp))
+    return;
+
+  auto elementType = nonWideningLinalgElementType(genericOp);
+  if (failed(elementType))
+    return;
+
+  if (elementType->isF32()) {
+    sizes.append({4, 4});
+    scalableSizeFlags.append({true, true});
+  }
+
+  if (elementType->isF64()) {
+    sizes.append({2, 2});
+    scalableSizeFlags.append({true, true});
+  }
```

I think that you also need to make sure that for element types other than `f32` and `f64`, the pre-condition fails. In the spirit of failing early.

---

_technically_ this fails no later than if it were in the pre-condition (but avoids duplicating it there, so I think this is fine).

---

As Ben points out, doing this would require duplicating the logic which isn't ideal, particularly now that the precondition isn't specific to SME, so going to leave this as is.